### PR TITLE
bpf: Add support for user tracers

### PR DESCRIPTION
**Pull request by alessandrogario**
demo: https://asciinema.org/a/419326

# Example usage

## Loading a tracer

```
sudo ./osquery/osqueryd \
  -S \
  --verbose \
  --disable_events=false \
  --enable_bpf_user_tracers=true \
  --bpf_user_tracer_list=tools/user_tracers/libcurl_tracer.json \
  --bpf_max_user_tracer_rows=1024
```

# Tracing a syscall (kprobe)

## Configuration file

```json
{
  "table_name": "kill_events",
  "function_name": "kill",
  "parameter_list": [
    {
      "name": "pid",
      "type": "Integer",
      "mode": "In",
      "size": 8
    },
    {
      "name": "signal",
      "type": "Integer",
      "mode": "In",
      "size": 8
    }
  ]
}
```

## Command

```
alessandro at ubuntu2104-rincewind in ~ 
$ killall -9 bash
```

## Table output

```
osquery> SELECT * FROM kill_events WHERE signal = 9;
+----------------+-----------+------------+------+------+-----------+-----------+-------------+----------+-------+--------+
| timestamp      | thread_id | process_id | uid  | gid  | cgroup_id | exit_code | probe_error | duration | pid   | signal |
+----------------+-----------+------------+------+------+-----------+-----------+-------------+----------+-------+--------+
| 28358125193419 | 29113     | 29113      | 1000 | 1000 | 9965      | 0         | 0           | 8179     | 29112 | 9      |
+----------------+-----------+------------+------+------+-----------+-----------+-------------+----------+-------+--------+
```

# Tracing libcurl (uprobe)

## Configuration file

```json
{
  "table_name": "libcurl_events",
  "path": "/lib/x86_64-linux-gnu/libcurl.so.4",
  "function_name": "curl_easy_setopt",
  "parameter_list": [
    {
      "name": "handle",
      "type": "Integer",
      "mode": "In",
      "size": 8
    },

    {
      "name": "option",
      "type": "Integer",
      "mode": "In",
      "size": 8
    },

    {
      "name": "value",
      "type": "String",
      "mode": "In"
    }
  ]
}
```

## Command

```
alessandro@ubuntu2104-rincewind:~$ curl https://google.com/
<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>301 Moved</TITLE></HEAD><BODY>
<H1>301 Moved</H1>
The document has moved
<A HREF="https://www.google.com/">here</A>.
</BODY></HTML>
```

## Table output

```
osquery> SELECT * FROM libcurl_events WHERE option = 10002;
+----------------+-----------+------------+------+------+-----------+-----------+-------------+----------+-------------+--------+---------------------+
| timestamp      | thread_id | process_id | uid  | gid  | cgroup_id | exit_code | probe_error | duration | handle      | option | value               |
+----------------+-----------+------------+------+------+-----------+-----------+-------------+----------+-------------+--------+---------------------+
| 27671658157992 | 27747     | 27747      | 1000 | 1000 | 9560      | 0         | 0           | 3717     | -1483881088 | 10002  | https://google.com/ |
+----------------+-----------+------------+------+------+-----------+-----------+-------------+----------+-------------+--------+---------------------+
```

# Tracing bash's readline (uprobe)

## Configuration file

```json
{
  "table_name": "libreadline_events",
  "path": "/usr/bin/bash",
  "function_name": "readline",
  "parameter_list": [
    {
      "name": "prompt",
      "type": "String",
      "mode": "In"
    },
    {
      "name": "EXIT_CODE",
      "type": "String",
      "mode": "Out"
    }
  ]
}
```

## Command

```
alessandro@ubuntu2104-rincewind:~$ date
Fri Apr 30 01:47:16 PM EDT 2021

alessandro@ubuntu2104-rincewind:~$ curl https://google.com
<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>301 Moved</TITLE></HEAD><BODY>
<H1>301 Moved</H1>
The document has moved
<A HREF="https://www.google.com/">here</A>.
</BODY></HTML>
```

## Table output

```
osquery> SELECT * FROM libreadline_events;
+----------------+-----------+------------+------+------+-----------+----------------+-------------+------------+-------------------------------------+-------------------------+
| timestamp      | thread_id | process_id | uid  | gid  | cgroup_id | exit_code      | probe_error | duration   | prompt                              | return_value            |
+----------------+-----------+------------+------+------+-----------+----------------+-------------+------------+-------------------------------------+-------------------------+
| 27911368786803 | 27063     | 27063      | 1000 | 1000 | 9560      | 94075622808240 | 0           | 1289583516 | alessandro@ubuntu2104-rincewind:~$  | date                    |
| 27929838950619 | 27063     | 27063      | 1000 | 1000 | 9560      | 94075622649856 | 0           | 1159185489 | alessandro@ubuntu2104-rincewind:~$  | curl https://google.com |
+----------------+-----------+------------+------+------+-----------+----------------+-------------+------------+-------------------------------------+-------------------------+
```

### CODE REVIEW

**Comment by zwass** [comment]
Looks like an awesome piece of functionality!

Can we make `process_id` be `pid` to match other osquery tables?

How tricky would it be to make this configurable at runtime (so it could be added to the config like atc)?

File: osquery/events/linux/bpf/usertracer.cpp [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+static const std::string kKprobeSyscallPrefix{"__x64_sys_"};
+
+#else
+#error Unsupported architecture
+#endif
+
+} // namespace
+
+FLAG(uint64,
+     max_bpf_user_tracer_rows,
```
Comment (theopolis) [inline osquery/events/linux/bpf/usertracer.cpp:37]: nitpick, `bpf_max_user_tracer_rows`?

File: osquery/events/linux/bpf/usertracer.cpp [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+  }
+
+  return config;
+}
+
+Expected<TableColumns, UserTracer::ErrorCode> UserTracer::generateTableSchema(
+    const Configuration& configuration) {
+  TableColumns column_list = {
+      std::make_tuple("timestamp", TEXT_TYPE, ColumnOptions::DEFAULT),
+      std::make_tuple("thread_id", INTEGER_TYPE, ColumnOptions::DEFAULT),
```
Comment (theopolis) [inline osquery/events/linux/bpf/usertracer.cpp:352]: nitpick, `tid`

File: osquery/events/linux/bpf/usertracer.cpp [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+      std::make_tuple("process_id", INTEGER_TYPE, ColumnOptions::DEFAULT),
```
Comment (theopolis) [inline osquery/events/linux/bpf/usertracer.cpp:353]: nitpick, `pid`

File: osquery/events/linux/bpf/usertracer.cpp [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+
+TableRows UserTracer::parseEvents(const EventQueue& event_queue) {
+  TableRows output;
+
+  for (const auto& event_queue_p : event_queue) {
+    const auto& event = event_queue_p.second;
+
+    auto row = make_table_row();
+    row["timestamp"] = SQL_TEXT(std::to_string(event.header.timestamp));
+    row["thread_id"] = INTEGER(event.header.thread_id);
```
Comment (theopolis) [inline osquery/events/linux/bpf/usertracer.cpp:457]: see nitpicks above

File: osquery/events/linux/bpf/usertracermanager.cpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+    auto status = readFile(path, tracer_config, 0);
+    if (!status.ok()) {
+      LOG(ERROR) << "Failed to load the user tracer configuration: " << path;
+      continue;
+    }
+
+    auto tracer_exp = UserTracer::create(tracer_config);
+    if (tracer_exp.isError()) {
+      LOG(ERROR) << "Failed to create the user tracer:"
```
Comment (theopolis) [inline osquery/events/linux/bpf/usertracermanager.cpp:56]: `: ` (missing space)

**Review by theopolis — DISMISSED** [review]
Pretty cool!

is there any way to test this from a unit test perspective?

**Comment by theopolis** [comment]
Could we also add a page to the documentation on how to use this before we land it?

**Comment by alessandrogario** [comment]
> Looks like an awesome piece of functionality!
> 
> Can we make `process_id` be `pid` to match other osquery tables?
> 
> How tricky would it be to make this configurable at runtime (so it could be added to the config like atc)?

The user tracers are now configured from the main config file, and will drop removed/outdated tracers and create new ones as the configuration is updated. The documentation has been added to the wiki: https://github.com/osquery/osquery/pull/7088/files?short_path=2b9d50b#diff-2b9d50b861c6cc9fc902807ce8d8aeb49f82709e83a27d48df9b8a27840e0298

**Comment by luc-lynx** [comment]
This looks like an awesome piece of functionality. Is there any chance to have this PR landed?

**Comment by alessandrogario** [comment]
> This looks like an awesome piece of functionality. Is there any chance to have this PR landed?

This is mostly an experiment, and I would like to see additional use cases before we merge it. What kind of events would you be interested in tracing with this functionality?

**Comment by grogsaxle** [comment]
Oh man!   I just found this.   I agree with @luc-lynx This is great stuff.   As for events that would be interested in tracing, I'm not sure the answer to that question matters.   maybe the answer is "all the things!" as this gives the security teams the flexibility to determine that themselves.   The libcurl demo is top notch for giving them ideas of the possible.

**Pull request closed without merging**